Task/136 main process refactor & test coverage

davemfish opened the pull request:
This PR aims to make the electron `main` process code more modular and testable, and more separated from code that runs in the `renderer` process. It fixes natcap/invest-workbench#136 and fixes natcap/invest-workbench#141 .

There is a new `src/main/` directory with all the code that runs in the `main` process. Those new files are probably a good place to start when reviewing.

There is also a new `tests/main/main.test.js`. Testing `main` process code required additional mocking  of the electron APIs, see `__mocks__/electron.js`.

This PR lays the groundwork for natcap/invest#870 by moving most of the logic that interacts with the filesystem (and most of the nodejs APIs in general) to the main process - out of the renderer. In particular, invest model subprocesses are now managed by main and invoked via IPC. There are still some lingering node APIs in the renderer (e.g. in `LogTab.jsx`) that will be dealt with in natcap/invest#870. This PR also adds a basic `preload.js` for sharing some APIs that need to be called from both `main` and `renderer` (e.g. logging and checking for devMode). More will need to go in there to accomplish natcap/invest#870 . 

Tests seem to be much more reliable now. A few race conditions were revealed during the refactor. And some event listeners in react components were previously not being cleaned up before the component unmounted. So that's fixed by using `componentWillUnmount` method where needed. 

Tests were re-organized a bit so now `npm run test` runs all tests except the `puppeteer` end-to-end test (`npm run test-electron-app`). 

Here's test coverage from `npm run test`. Some of the main process modules that are still low coverage tend to be very thin wrappers around boilerplate electron APIs.

```
----------File             | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s ---------------------------|---------|---
 src                       |      85 |    60.87 |   69.12 |   84.91 |
  InvestJob.js             |     100 |    93.75 |     100 |     100 | 24
  app.jsx                  |     100 |      100 |     100 |     100 |
  server_requests.js       |   65.38 |      100 |   60.87 |   65.38 | ...,79-97,121,145
  ui_config.js             |   21.43 |    11.11 |    8.33 |   21.43 | 38,73-404
  utils.js                 |   71.43 |       50 |      75 |   71.43 | 19,45-47
 ...ents/DataDownloadModal |     100 |       90 |     100 |     100 |
  index.jsx                |     100 |       90 |     100 |     100 | 56
 src/components/Expire     |     100 |      100 |     100 |     100 |
  index.jsx                |     100 |      100 |     100 |     100 |
 src/components/HomeTab    |     100 |      100 |     100 |     100 |
  index.jsx                |     100 |      100 |     100 |     100 |
 src/components/InvestTab  |   94.83 |       80 |     100 |   94.83 |
  index.jsx                |   94.83 |       80 |     100 |   94.83 | 37-41
 src/components/LoadButton |     100 |      100 |     100 |     100 |
  index.jsx                |     100 |      100 |     100 |     100 |
 src/components/LogTab     |   88.41 |    91.67 |   85.71 |   88.41 |
  index.jsx                |   88.41 |    91.67 |   85.71 |   88.41 | ...58-165,175,224
 src/components/Portal     |     100 |      100 |     100 |     100 |
  index.jsx                |     100 |      100 |     100 |     100 |
 ...ponents/ResourcesLinks |   88.24 |       75 |      50 |   88.24 |
  index.jsx                |   88.24 |       75 |      50 |   88.24 | 38-39
 ...ponents/SaveFileButton |     100 |      100 |     100 |     100 |
  index.jsx                |     100 |      100 |     100 |     100 |
 ...mponents/SettingsModal |   95.08 |      100 |     100 |   95.08 |
  SettingsStorage.js       |    87.5 |      100 |     100 |    87.5 | 57-58,78
  index.jsx                |     100 |      100 |     100 |     100 |
 src/components/SetupTab   |     100 |      100 |     100 |     100 |
  index.jsx                |     100 |      100 |     100 |     100 |
 ...ents/SetupTab/ArgInput |      88 |    86.67 |   84.62 |      88 |
  index.jsx                |      88 |    86.67 |   84.62 |      88 | 65-70,318
 ...ents/SetupTab/ArgsForm |   88.89 |    61.11 |   91.67 |   88.89 |
  index.jsx                |   88.89 |    61.11 |   91.67 |   88.89 | ...44-45,52,93,97
 .../SetupTab/SetupButtons |     100 |      100 |     100 |     100 |
  index.jsx                |     100 |      100 |     100 |     100 |
 src/main                  |   63.67 |    36.78 |   61.43 |   64.55 |
  ...PythonFlaskProcess.js |   65.79 |    42.86 |   66.67 |   64.86 | ...6-59,87-92,110
  extractZipInplace.js     |      70 |    28.57 |   88.89 |   80.77 | 24-28,41
  findInvestBinaries.js    |     100 |       75 |     100 |     100 | 17
  ipcMainChannels.js       |     100 |      100 |     100 |     100 |
  main.js                  |   50.91 |       20 |   36.36 |   50.91 | ...17,133-164,170
  menubar.js               |   29.41 |    33.33 |      20 |   29.41 | 96-100,109-144
  setupCheckFirstRun.js    |   83.33 |      100 |   66.67 |   83.33 | 29,36
  setupContextMenu.js      |      20 |      100 |   33.33 |      20 | 11-21
  setupDialogs.js          |   33.33 |      100 |   33.33 |   33.33 | 11-12,18-19
  setupDownloadHandlers.js |   28.21 |        0 |    37.5 |   27.78 | 30-67
  setupInvestHandlers.js   |   96.72 |       75 |     100 |   96.61 | 71,116
---------------------------|---------|----------|---------|---------|-------------------Test Suites: 10 passed, 10 total
Tests:       82 passed, 82 total
```

davemfish committed:
treat main.js as a module so we can import it in tests. #136

davemfish committed:
preload babel/register on npm start to use es6 module syntax in main.js. also a better dev-mode checker. #136.

davemfish committed:
added module to setup the ipcMain handlers. #136.

davemfish committed:
moved main process code to a main folder. #136.

davemfish committed:
fixing some relative paths after the main folder move. #136.

davemfish committed:
added separate modules for other ipcMain handlers. #136

davemfish committed:
fixed relative path import

davemfish committed:
moved download handlers to their own module. #136.

davemfish committed:
deleting some comments. #136.

davemfish committed:
added unit test for checkFirstRun. 136.

davemfish committed:
working on integration test for the createWindow function. #136.

davemfish committed:
needed some more mocking for createWindow test. #136

davemfish committed:
more mocks of the electron main process APIs. #136.

davemfish committed:
removed a redundant test. #136.

davemfish committed:
unit tests for findInvestBinaries, now in its own module and now not returning a Promise. #136.

davemfish committed:
removed an unused variable in findInvestBinaries. #136.

davemfish committed:
changed npm run test to run all test suites except e2e tests. #136.

davemfish committed:
started to setup ipc communication for spawning invest from main process. #136.

davemfish committed:
lots of cleanup following the previous commit's changes. #136.

davemfish committed:
patching some tests that use argsToJsonFile. #136

davemfish committed:
removed an unecessary ipc calls for writing invest params to json. #136.

davemfish committed:
removed the constants.js module and its fileRegistry object. #136.

davemfish committed:
fixed bug in creating the temp datastack to pass to cli. #136.

davemfish committed:
debugged the string we use as a unique IPC channel for each invest job. #136.

davemfish committed:
developed IPC mocks to work with invest subprocess tests. in progress. #136.

davemfish committed:
invest subprocess tests passing with functional ipc send, on, mocks. #136

davemfish committed:
moved download modal integration tests to main.test because they depend on the BrowserWindow instance in main's scope. #136.

davemfish committed:
added unit test for extractZipInplace. #136

davemfish committed:
don't need logging from this page. #136.

davemfish committed:
finished cleanup for extractzip test. #136.

davemfish committed:
switched to electron-log & added a preload script to access log api from renderer. #136 and #141.

davemfish committed:
added mock for electron-log and further separated renderer & main process code so that no modules are imported by both processes. #136. #141.

davemfish committed:
avoiding a server_requests import from main. #136.

davemfish committed:
removed some unused ipcMain handlers. #136.

davemfish committed:
fetch invest binaries for GHA run-tests workflow. #136.

davemfish committed:
moved flask app tests out of GHA build workflow and into tests workflow. #136.

davemfish committed:
making sure to remove ipcMain event listeners on shutdown. #136.

davemfish committed:
cleaning up ipcRenderer event listeners. #136.

davemfish committed:
debugging a mkdir on macos GHA. #136.

davemfish committed:
improper use of path.resolve. #136.

davemfish committed:
debugged build config given new location of main.js. #136.

davemfish committed:
updated build script to avoid using spawn to run babel. #136.

davemfish committed:
fixing a race condition in mkdirs. #136

davemfish committed:
making sure to do test setup inside a beforeEach. #136.

davemfish committed:
using sync fs calls. #136.

davemfish committed:
debugging a race condition with fs on mac GHA. #136.

davemfish committed:
still need to throw after catching error. #136.

davemfish committed:
updated invest version to avoid the bug with logging from ui_server.py #136.

davemfish committed:
do not start invest server with --debug. #136.

davemfish committed:
updated path to electron's logs so they continue to upload as github artifacts. #136.

davemfish committed:
fixed build script to allow fetching invest from releases or dev builds bucket. #136.

davemfish committed:
removed the mac-related instance lock logic because we're only taking that path on Windows. #136.

davemfish committed:
a couple comments cleaned up. #136.

davemfish committed:
debugging race condition in zip extract test. #136.

davemfish committed:
trying to reproduce a race condition with a more minimal test. #136.

davemfish committed:
installed rimraf to replace deprecrated fs recursive rmdir. #136.

davemfish committed:
a couple more applications for rimraf. #136.

davemfish committed:
still debugging macos race condition - move test setup & teardown out of beforeEach, into test. #136.

davemfish committed:
still debugging macos race condition. #136.

davemfish committed:
still debugging macos race condition. #136.

davemfish committed:
still debugging macos race condition with many console.logs. #136.

davemfish committed:
still debugging macos race condition - waiting for all the setup expects. #136.

davemfish committed:
still debugging macos race condition - some more logic should have been in a callback. #136.

davemfish committed:
race condition seems fixed by previous commit, removing console.logs. #136.

davemfish committed:
putting setup and teardown back in before/afterEach. #136.

davemfish committed:
do not reject from a promise unless we really need an async rejection. most of the time we want to throw. #136.

davemfish committed:
cleaned up the electron IPC mocks. #136.

davemfish committed:
some more cleanup in the electron mocks. #136.

davemfish committed:
removed all uses of dotenv package. the costs of maintaining the extra logic outweigh the benefit of using this package to ease local development. #136.

davemfish committed:
cleaning up a docstring. #136.

davemfish committed:
uninstalled an unused dependency. #136.

davemfish committed:
cleaned up unused parts of the build script. #136.

davemfish committed:
moved invest binaries back to HEAD of release/workbench-alpha on natcap's fork. #136.

davemfish committed:
remove some unused npm scripts. #136.

davemfish committed:
revert unneeded changes to babel config. #136.

davemfish committed:
removed some commented code. #136.

davemfish committed:
removed some commented code. #136.

emlys commented on src/main/setupDownloadHandlers.js:
```diff
@@ -0,0 +1,71 @@
+
+const logger = getLogger(__filename.split('/').slice(-1)[0]);
+
+export default function setupDownloadHandlers(mainWindow) {
```
need a docstring here?

davemfish replied:
Thanks, added a docstring. 

> If I'm understanding correctly, it's ok to use the mainWindow object directly here, because this function is called from main.js and so it's run in the main process? Whereas things running in the renderer process need to use IPC?

Yes, I think that's right. Though to be honest I'm not sure what would happen if renderer process function was passed a reference to `mainWindow`. 

Anyway, all of this was previously inside the `createWindow` function in `main.js` where `mainWindow` was already in scope. But that was pretty messy so I wanted a module devoted just to this stuff. But I also needed to use `mainWindow.webContents.send`, so that's why `mainWindow` is passed to the function. I'm not certain that is a good idea though - passing around `mainWindow`. Any thoughts?

emlys replied:
This seems fine to me

emlys commented on src/main/setupDownloadHandlers.js:
```diff
@@ -0,0 +1,71 @@
+    });
+
+  mainWindow.webContents.session.on('will-download', (event, item) => {
+    const filename = item.getFilename();
```
maybe worth a comment that `item` is a `DownloadItem` object

emlys commented on src/main/setupDownloadHandlers.js:
```diff
@@ -0,0 +1,71 @@
+        if (item.isPaused()) {
+          logger.info('download paused');
+        } else {
+          logger.info(`${item.getSavePath()}`);
```
maybe add context like `save path: ${item.getSavePath()}`

emlys reviewed (changes requested):
Hey @davemfish, I'm posting what I have reviewed so far so hopefully me and Doug don't overlap too much. It's great having lots of the functions separated into different files in src/main, and what I've looked at is a lot cleaner and more understandable than I remember from before!

davemfish committed:
no need to validate the argument inside this function. #136.

davemfish commented on src/main/createPythonFlaskProcess.js:
```diff
@@ -0,0 +1,112 @@
+      logger.debug(`Flask process disconnected`);
+    });
+  } else {
+    logger.error('no existing invest installations found');
```
You're right, at this point we already know we have the path to a valid invest exe, so I'm taking that check out of this function altogether.

davemfish commented on src/main/createPythonFlaskProcess.js:
```diff
@@ -0,0 +1,112 @@
+
+/** Find out if the Flask server is online, waiting until it is.
+ *
+ * Sometimes the app will make a server request before it's ready,
```
That's exactly right. The flask server takes a couple seconds to startup, so we need to wait for it. But you're right, this is a better comment for where the function is called than the docstring.

davemfish commented on src/main/createPythonFlaskProcess.js:
```diff
@@ -0,0 +1,112 @@
+      method: 'get',
+    })
+      .then((response) => response.text())
+      .catch(async (error) => {
```
This is promise chaining and error propagation at work! https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#error_propagation

emlys replied:
Ohh okay, so the `.catch` will catch any errors that happen above it in the chain. Thanks for the link

davemfish committed:
removed some unnecessary docstring bits. #136.

davemfish commented on src/main/createPythonFlaskProcess.js:
```diff
@@ -0,0 +1,112 @@
+          throw error;
+        } else {
+          logger.error(error.stack);
+          throw error;
```
Well we're expecting to catch an 'ECONNREFUSED' for the first few fetch tries, and we want to wait & retry when that happens. So without the `else` we would `throw` on that error, which isn't what we want, right? We just want to throw on other unexpected errors.

emlys replied:
I think it would still retry? Whether or not we take the `if` path or the `else` path, the lines `         logger.error(error.stack);` and `throw error;` are always called at the end (unless we've already `return`ed), so I think this is equivalent:
```
if (error.code === 'ECONNREFUSED') {
  while (i < retries) {
    i++;
    // Try every X ms, usually takes a couple seconds to startup.
    await new Promise((resolve) => setTimeout(resolve, 300));
    logger.debug(`retry # ${i}`);
    return await getFlaskIsReady({ i: i, retries: retries });
  }
  logger.error(`Not able to connect to server after ${retries} tries.`);
}
logger.error(error.stack);
throw error;
```
If the error is an 'ECONNREFUSED', we retry. If we're out of retries, we log the "Not able to connect to server" message, then log the error stack and throw the error. If the error is something other than 'ECONNREFUSED', we just log the error stack and throw the error. Let me know if I'm missing something!

davemfish replied:
Oh yes, thank you! I was ignoring the return statement.

davemfish commented on src/main/extractZipInplace.js:
```diff
@@ -0,0 +1,55 @@
+import fs from 'fs';
+import path from 'path';
+
+import yauzl from 'yauzl';
```
Haha I never actually looked up the acronym. But I did read at least one stackoverflow post before choosing the library. https://stackoverflow.com/questions/10308110/simplest-way-to-download-and-unzip-files-in-node-js-cross-platform

davemfish commented on src/main/extractZipInplace.js:
```diff
@@ -0,0 +1,55 @@
+
+const logger = getLogger(__filename.split('/').slice(-1)[0]);
+
+export default function extractZipInplace(zipFilePath) {
```
Thank you!

davemfish committed:
added docstring for extractZip function. #136.

davemfish committed:
added comment about unzip options. #136.

davemfish commented on src/main/extractZipInplace.js:
```diff
@@ -0,0 +1,55 @@
+    logger.info(`extracting ${zipFilePath}`);
+    const options = {
+      lazyEntries: true,
+      autoClose: true,
```
Thanks! I removed that default option and added a comment for `lazyEntries`

davemfish commented on src/main/extractZipInplace.js:
```diff
@@ -0,0 +1,55 @@
+        if (/\/$/.test(entry.fileName)) {
+          fs.mkdir(writePath, (e) => {
+            if (e) {
+              if (e.code === 'EEXIST') { } else throw e;
```
Oh yes, much better!

davemfish commented on src/main/extractZipInplace.js:
```diff
@@ -0,0 +1,55 @@
+      zipfile.on('close', () => {
+        resolve(true);
+      });
+      zipfile.readEntry();
```
I agree. I know such a thing is not in `yauzl`. It might exist elsewhere and I'm open to switching. Programming this taught me a bit about read/writeStreams and promises, so no regrets!

For our use case, we also want to make sure the dir structure of the output matches the zip archive, which may or may not be true for some higher-level library. Especially because I think sometimes the directories themselves are "entries" in the archive and sometimes they are not (even when the archived files have the dir in their path). So it took a bit of custom logic to make sure the output dir is always created.

emlys replied:
I read that thread you linked above and it's too bad there isn't one stable, generally-agreed-upon unzip npm module. Could you remind me what we're downloading? When would the output structure not match the archive? If you can accomplish this with `unzip` or another shell command, maybe calling that from `child_process` would be more straightfoward

davemfish replied:
Yep, these are the invest sample data downloads. So they're folders with varying levels of subfolders. 

`unzip` would be nice, but I hesitate to depend on a program that we're not actually distributing with the app. Windows users won't have `unzip`, for example.

davemfish replied:
@emlys to see an example archive that would not necessarily extract to the same output structure without all this logic, check out the test we have for this, where the directories themselves are _not_ "entries" in the archive. https://github.com/natcap/invest-workbench/pull/147/files#diff-84796f464c3e43389936aa12df7ce439dab4dc511f7bc43e7143cd03be25197aR107

Of course if we do find a nice higher-level utility for this, I would hope it handles this case by default. I assume `unzip` does.

davemfish commented on src/main/ipcMainChannels.js:
```diff
@@ -0,0 +1,9 @@
+  INVEST_RUN: 'invest-run',
+  INVEST_KILL: 'invest-kill',
+  SHOW_CONTEXT_MENU: 'show-context-menu',
+  IS_FIRST_RUN: 'is-first-run',
```
Yes, that's true. But there will likely be a few more to come when we fully tackle #60

davemfish commented on src/main/main.js:
```diff
@@ -0,0 +1,174 @@
+  nativeTheme,
+  Menu,
+  ipcMain,
+} from 'electron'; // eslint-disable-line import/no-extraneous-dependencies
```
This warning is actually about electron being absent from package.json `dependencies`. It's under `devDependencies` instead. Eslint thinks that's wrong since we're importing it here, but electron-builder wants it only listed in devDeps. Otherwise I think we'd probably end up distributing two electron modules in production.

emlys replied:
Oh weird, thanks for the explanation.

davemfish commented on src/main/main.js:
```diff
@@ -0,0 +1,174 @@
+process.env.PORT = '56789';
+
+// Keep a global reference of the window object, if you don't, the window will
+// be closed automatically when the JavaScript object is garbage collected.
```
This one traces all the way back to the tutorials/docs. Thanks electron maintainers!

davemfish commented on src/main/main.js:
```diff
@@ -0,0 +1,174 @@
+  app.on('ready', createWindow);
+  app.on('activate', () => {
+    if (mainWindow === null) {
+      createWindow();
```
I think this is what happens on MacOS if you have previously closed the window but the application is still running, and you open it again?

emlys replied:
Makes sense!

davemfish commented on src/main/main.js:
```diff
@@ -0,0 +1,174 @@
+  });
+}
+
+if (typeof require !== 'undefined' && require.main === module) {
```
That's the equivalent of `if __name__ == "__main__":`

So it let's us import the file as a module, just for tests, without immediately calling `main()`.

davemfish commented on src/main/main.js:
```diff
@@ -0,0 +1,174 @@
+  main(process.argv);
+}
+
+// module.exports = main;
```
Oh, I guess we're not actually ever importing `main` function in any tests. Yes I'll remove that.

davemfish committed:
fixing error handling syntax. #136.

davemfish committed:
removed some commented code. #136.

davemfish commented on src/main/setupCheckFirstRun.js:
```diff
@@ -0,0 +1,38 @@
+ *
+ * @returns {boolean}
+ */
+export function checkFirstRun() {
```
We're importing it in tests. I don't love adding implementation details only for the sake of tests, but I've gotten over it for stuff like this 😄

davemfish commented on src/main/setupCheckFirstRun.js:
```diff
@@ -0,0 +1,38 @@
+  return true;
+}
+
+export function setupCheckFirstRun() {
```
I dunno, it's such a thin wrapper around the electron API call. I'm okay without a docstring, but that's just one opinion.

emlys replied:
That's fine by me

davemfish commented on src/main/setupContextMenu.js:
```diff
@@ -0,0 +1,23 @@
+      },
+    ];
+    const menu = Menu.buildFromTemplate(template);
+    menu.popup(BrowserWindow.fromWebContents(event.sender));
```
Agreed. Looks like I only added it so we could have an Inspect Element item. Super useful for us. Maybe we'll want the standard copy/paste stuff in there instead for production.

davemfish committed:
added a docstring to setupDownloadHandlers. #136.

davemfish commented on src/main/setupDownloadHandlers.js:
```diff
@@ -0,0 +1,71 @@
+  let downloadDir;
+  let downloadLength;
+  const downloadQueue = [];
+  ipcMain.on(ipcMainChannels.DOWNLOAD_URL,
```
My understanding is that `invoke-handle` pairs are ideal for when a value needs to be returned. So it's used when we need to ask the main process for some quick data. 

And `send-on` pairs are better for triggering other longer running side-effects to run in main, but not returning data from main back to renderer.

I think the docs are better for the ipcRenderer methods vs ipcMain for some reason. https://www.electronjs.org/docs/api/ipc-renderer#ipcrenderersendchannel-args

And then `handleOnce` and `once` (vs. `handle` and `on`) automatically remove the event listener after they've handled one event. So they aren't always waiting & listening for more signals.

davemfish commented on src/main/setupDownloadHandlers.js:
```diff
@@ -0,0 +1,71 @@
+      downloadLength = downloadQueue.length;
+      mainWindow.webContents.send(
+        'download-status',
+        [(downloadLength - downloadQueue.length), downloadLength]
```
Right, they are the same here when we first handle the event from the renderer with the array of URLs to download. But then later when we're handling each individual 'will-download' event, we remove items from `downloadQueue` as we go,  but `downloadLength` remains the original queue length.

emlys replied:
Ah okay, I see it's consistent with the other call to `mainWindow.webContents.send` below. On this line, `downloadLength - downloadQueue.length` could be replaced with `0` and do the same thing, though? (Not saying you need to change it, just making sure I understand)

davemfish replied:
Right, I think it is more readable to use `0` there, thanks.

davemfish committed:
added a comment. #136.

davemfish committed:
added some context to some logger messages. #136.

davemfish commented:
> Hey @davemfish, I'm posting what I have reviewed so far so hopefully me and Doug don't overlap too much. It's great having lots of the functions separated into different files in src/main, and what I've looked at is a lot cleaner and more understandable than I remember from before!

Thanks for the positive feedback, @emlys ! I believe I addressed all your comments so far and pushed some changes.

davemfish committed:
not using handleOnce for 'is-first-run' so that we can restart the renderer (browser refresh) without also restarting the main. #136.

davemfish committed:
fixed bug in a function signature. #136

davemfish committed:
there's no longer any reason for this function to be async. #136.

emlys commented on src/main/setupCheckFirstRun.js:
```diff
@@ -0,0 +1,38 @@
+
+export function setupCheckFirstRun() {
+  ipcMain.handleOnce(
+    ipcMainChannels.IS_FIRST_RUN, (event) => checkFirstRun()
```
I found [this answer](https://stackoverflow.com/a/47084018/14451410) which says you can call any javascript function with any number of arguments. So I imagine this is fine to stay as it is. If you want to change them to `() => checkFirstRun()`, that would make sense, since no arguments are being used.

davemfish replied:
Cool, thanks for the reference! I think I'll change them to `() => { }` since having an unused argument seems undesirable too. Especially if this were a much longer function, it wouldn't immediately be obvious that `event` was not used.

emlys commented on .github/workflows/build-electron.yml:
```diff
@@ -94,10 +94,6 @@ jobs:
 
-      # Also run all tests on the build dir code?
-    - name: Test flask app binaries
-      run: npm run test-flask-app
```
Do we still want this deleted?

davemfish replied:
Yeah these tests now run as part of `npm run test`.

dcdenu4 commented on src/main/extractZipInplace.js:
```diff
@@ -0,0 +1,64 @@
+            // Sometimes an entry will be in a dir, where the
+            // dir itself was *not* an entry, therefore we still need
+            // to create the dir here.
+            fs.mkdir(path.dirname(writePath), (e) => {
```
Just going to see if I've got a handle on what `fs.mkdir` is doing here.
`writePath` could be a file name in a dir structure that hasn't been made yet, so `mkdir` will make that directory chain if it is NOT already there and then go ahead and write the file.

davemfish replied:
Oh but there is a bug here because `mkdir` does not make all the directories up the chain and there's no guarantee they all exist.

davemfish committed:
simplified the error catching logic in getFlaskIsReady. #136.

davemfish commented on src/main/setupContextMenu.js:
```diff
@@ -0,0 +1,23 @@
+import { ipcMainChannels } from './ipcMainChannels';
+
+export default function setupContextMenu() {
+  ipcMain.on(ipcMainChannels.SHOW_CONTEXT_MENU, (event, rightClickPos) => {
```
Leaving this as-is because we're using those both those args.

davemfish commented on src/main/setupDialogs.js:
```diff
@@ -0,0 +1,22 @@
+  );
+
+  ipcMain.handle(
+    ipcMainChannels.SHOW_SAVE_DIALOG, async (event, options) => {
```
Here we're not using `event`, but it's expected as the first argument so I think it needs to stay in order pass `options`.

davemfish committed:
removed an unused event argument. #136

davemfish committed:
simplifying something in downloadHandlers function. #136

davemfish committed:
fixed bug in extractZip - it needs to do a recursive mkdir if necessary. updated the test case to cover that case. #136.

davemfish committed:
forgot to remove a variable reference. #136.

davemfish committed:
fixed initial props passed to App in tests and in initial render. #136.

davemfish committed:
updated to the latest actions/upload-artifact. #136.

davemfish committed:
moved pptr screenshots in the logs dir created by electron-log - for more convenient github artifacts uploads. #136.

davemfish committed:
fixing the path to electron-log's default location on macos. #136.

emlys commented on src/sampledata_registry.json:
```diff
@@ -1,133 +1,133 @@
-    "url": "https://www.googleapis.com/download/storage/v1/b/releases.naturalcapitalproject.org/o/invest%2F3.9.0.post285%2Bgf68dd440%2Fdata%2FCarbon.zip?generation=1614051202294534&alt=media",
-    "filesize": "1900216"
+    "url": "https://www.googleapis.com/download/storage/v1/b/releases.naturalcapitalproject.org/o/invest%2F3.9.0.post432%2Bgdecbeffb%2Fdata%2FCarbon.zip?generation=1622754766634004&alt=media",
+    "filesize": "1900623"
```
Curious why we need to record the file size for all these?

davemfish replied:
We're displaying it to the user in the download component, in units of MB. I think Doug reviewed that PR so you may not have seen that feature yet!

emlys commented on tests/app.test.js:
```diff
@@ -458,9 +462,21 @@ describe('InVEST subprocess testing', () => {
   let fakeWorkspace;
   let logfilePath;
   let mockInvestProc;
+  let spyKill;
```
such a dramatic variable name! 🔫

emlys commented on __mocks__/electron.js:
```diff
@@ -1,22 +1,80 @@
+
+APIs are only available in an electron browser environment, but our tests
+run in node where all electron's exports are undefined. So we must mock,
+even for APIs that would otherwise seem okay to call during a test.
```
Could this file go in the test directory? I'm not sure about having a __mocks__ directory at the top level

davemfish replied:
It actually needs to go at the same level as `node_modules` in order to automatically mock the node module in all the tests. There are other possible ways to load mocks, but this seemed like a good choice for mocks that always need to be loaded for all tests.

There's a similar setup at `src/__mocks__/logger.js` to always mock `src/logger.js`.

emlys reviewed (commented):
This looks good Dave! There's still a lot I don't fully understand, but for the goal of refactoring the main process and improving tests, this is great. It's good to see that the integration tests have been reliably passing (or failing for an obvious reason) in the recent GHA runs. I just had some more minor questions. If you and @dcdenu4 finish this up while I'm gone next week, feel free to go ahead and merge!

davemfish commented on build.js:
```diff
@@ -1,14 +1,11 @@
 const BUILD_DIR = 'build';
-const ELECTRON_BUILDER_ENV = 'electron-builder.env';
 
 if (process.argv[2] && process.argv[2] === 'clean') {
```
yeah I think you're right. The pattern should probably be `if (process.argv && ....` but I think there will always be an `argv` property so that's excessive too.

davemfish commented on build.js:
```diff
@@ -35,38 +35,9 @@ function clean() {
-    files.forEach((file) => {
-      fs.unlinkSync(file);
-    });
-    fs.unlinkSync(ELECTRON_BUILDER_ENV);
```
I think we stopped using that env file sometime in the past in favor of `electron-builder-config.js` . Unfortunately this was still left here in the build script.

davemfish commented on scripts/fetch_invest_binaries.js:
```diff
@@ -22,20 +22,32 @@ switch (process.platform) {
+  binaryZipPath = `${bucket}/${repo}/${VERSION}/${binaryZipName}`;
+  DATA_PREFIX = `${repo}/${VERSION}/data`;
+} else if (bucket === 'natcap-dev-build-artifacts') {
+  const { fork } = pkg.invest;
```
Nope, that's better!

davemfish commented on scripts/fetch_invest_binaries.js:
```diff
@@ -22,20 +22,32 @@ switch (process.platform) {
+  const { fork } = pkg.invest;
+  binaryZipPath = `${bucket}/${repo}/${fork}/${VERSION}/${binaryZipName}`;
+  DATA_PREFIX = `${repo}/${fork}/${VERSION}/data`;
+}
```
Not knowing what to call that new variable, I'm just leaving it as-is. A bit repetitive, but readable, I think.

davemfish commented on scripts/fetch_invest_binaries.js:
```diff
@@ -109,17 +119,19 @@ async function updateSampledataRegistry() {
-
-  const dataItems = await queryStorage(dataEndpoint);
+  const dataItems = await queryStorage(DATA_QUERY_URL);
   Object.keys(registry).forEach((model) => {
```
Hmm, but we're also using the key again to reassign `registry[model].url`

davemfish commented on src/components/InvestTab/index.jsx:
```diff
@@ -152,111 +117,52 @@ export default class InvestTab extends React.Component {
+      'invest-run',
+      job.metadata.modelRunName,
+      this.state.modelSpec.module,
+      args,
```
That's right. Most of the stuff that was removed from here in `investExecute` is now in `main/setupInvestHandlers.js`. And `argsToJsonFile in particular is something that writes to the local filesystem, so is best handled by Node in the main process.

davemfish commented on src/components/LogTab/index.jsx:
```diff
@@ -200,8 +198,7 @@ export default class LogTab extends React.Component {
         variant="outline-dark"
-        onClick={() => terminateInvestProcess(procID)}
-        disabled={!procID}
+        onClick={terminateInvestProcess}
```
Yes, but we also conditionally render `CancelButton` only while the process is running. So the `disabled` property seemed extraneous.

davemfish committed:
minor changes in build script, for simplicity. #136.

dcdenu4 reviewed (approved):
This looks great, I'm excited to dive into this more in the future Dave!
